network/src/testutils/fake_socket.rs

```diff
@@ -0,0 +1,317 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//!
+//! This module exposes two types of sockets useful for tests:
+// - ExposingTestSocket: a wrapper around MemorySocket to retrieve handshake messages being sent.
+// - ReadOnlyTestSocket: a socket that can be read from in different ways.
+//!
+//! What is this fake socket for?
+//!
+//! - does nothing on a write
+//! - read behaves according to mode
+//!
+//!
+//! used to quickly consume fuzzing data and dismiss socket writes.
+//! ```
+//!
+//! ```
+
+use futures::{
+    io::{AsyncRead, AsyncWrite},
+    ready,
+    task::{Context, Poll},
+};
+use memsocket::MemorySocket;
+use std::{io, pin::Pin};
+
+//
+// ReadOnlyTestSocket
+// ==================
+//
+
+pub struct ReadOnlyTestSocket<'a> {
+    /// the content
+    content: &'a [u8],
+    /// the socket will read byte-by-byte
+    fragmented: bool,
+    /// continue to read 0s once content has been fully read
+    trailing: bool,
+}
+
+impl<'a> ReadOnlyTestSocket<'a> {
+    pub fn new(content: &'a [u8]) -> Self {
+        Self {
+            content,
+            fragmented: false,
+            trailing: false,
+        }
+    }
+
+    /// reads will have to be done byte by byte
+    #[allow(dead_code)]
+    pub fn set_fragmented_read(&mut self) {
+        self.fragmented = true;
+    }
+
+    /// reads will never return pending, but 0s
+    #[allow(dead_code)]
+    pub fn set_trailing(&mut self) {
+        self.trailing = true;
+    }
+}
+
+/// Does nothing, but looks to the caller as if write worked
+impl<'a> AsyncWrite for ReadOnlyTestSocket<'a> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        _context: &mut Context,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        Poll::Ready(Ok(buf.len()))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _context: &mut Context) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+
+    /// Attempt to close the channel. Cannot Fail.
+    fn poll_close(self: Pin<&mut Self>, _context: &mut Context) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// Read based on the mode set
+impl<'a> AsyncRead for ReadOnlyTestSocket<'a> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        mut _context: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        // nothing left to read
+        if self.content.is_empty() {
+            if self.trailing {
+                // [0, 0, 0, 0, ...]
+                for val in buf.iter_mut() {
+                    *val = 0;
+                }
+                return Poll::Ready(Ok(buf.len()));
+            } else {
+                // EOF
+                return Poll::Ready(Ok(0));
+            }
+        }
+
+        // if something left return what's asked
+        if self.fragmented {
```

Comment:
So fragmented will always be exactly 1 byte?  I think it's a fine estimation, just wondering if you'd want to put some variability in it.

Reply:
I thought about it, and it'd be easy to put some variability in it, but I don't think it's necessary. We can always add that later if we need it : o